File: 0042. Trapping Rain Water/Solution.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        res = left = pre_max = suf_max = 0
        right = len(height) - 1
        while left < right:
            pre_max, suf_max = max(pre_max, height[left]), max(suf_max, height[right])
            if pre_max < suf_max:
                res += pre_max - height[left]
                left += 1
            else:
                res += suf_max - height[right]
                right -= 1
        return res

File: 0042. Trapping Rain Water/Solution.c
#include <math.h>

int trap(int* height, int heightSize)
{
    int res = 0, pre_max = 0, suf_max = 0, i = 0, j = heightSize - 1;
    while (i < j) {
        pre_max = fmax(pre_max, height[i]);
        suf_max = fmax(suf_max, height[j]);
        res += pre_max < suf_max ? pre_max - height[i++] : suf_max - height[j--];
    }
    return res;
}
